k2/python/k2/fsa.py

```diff
@@ -227,92 +251,156 @@ def __setattr__(self, name: str, value: Any) -> None:
                 # to integer patterns here.
                 self.arcs.values()[:, -1] = _as_int(value.detach())
         else:
-            assert name != 'properties'
+            assert name != 'properties'  # should be set by getter/setter or by
+                                         # doing self.__dict__['_properties'] =
+                                         # foo
             self._non_tensor_attr[name] = value
 
-    def __getattr__(self, name: str) -> Any:
-        if name == 'labels':
+
+    @property
+    def labels(self) -> torch.Tensor:
+        try:
             return self.arcs.values()[:, 2]
-        elif name in self._tensor_attr:
+        except:
+            # print the exception because it will probably be lost, since
+            # python's getting code will back off to __getattr__.
+            traceback.print_exc()
+            raise e
+
+    @property
+    def labels(self, values) -> None:
+        assert value.dtype == torch.int32
+        self.arcs.values()[:, 2] = values
+        # Invalidate the properties since we changed the labels.
+        self._properties = None
+
+
+    @property
+    def properties(self) -> int:
+        # instead of accessing self._properties, we use
+        # self.__dict__.{get,set}('_properties') in order to
+        # avoid calling __getattr__ and any complexity involved in that.
+        properties = self.__dict__.get('_properties', None)
+        if properties != None:
+            return properties  # Return cached properties.
+
+        if self.arcs.num_axes() == 2:
+            properties = _k2.get_fsa_basic_properties(self.arcs)
+        else:
+            properties = _k2.get_fsa_vec_basic_properties(self.arcs)
+        self.__dict__['_properties'] = properties
+        if properties & fsa_properties.VALID != 1:
+            raise ValueError(
+                "Fsa is not valid, properties are: {} = {}, arcs are: {}".
+                format(properties, fsa_properties.to_str(properties),
+                       str(self.arcs)))
+        return properties
+
+    @properties.setter
+    def properties(self, value) -> None:
+        """ Only supports setting self.property to None; any other
+        value should be set internally by writing to self.__dict__ directly."""
+        if value == None:
+            self.__dict__['_properties'] = None
+        else:
+            raise RuntimeError("""Currently we don't allow the .properties of an Fsa
+            to be set this way, except to None""");
+
+    @property
+    def properties_str(self) -> str:
+        return _k2.fsa_properties_as_str(self._properties)
+
+    @property
+    def requires_grad(self) -> bool:
+        return self.scores.requires_grad
+
+    @property
+    def grad(self) -> bool:
```

Comment:
why is it a bool?

Reply:
I think this must be a  typo introduced by copying and pasting `def requires_grad`.